File: sdk/typescript/src/http.ts
/**
 * HTTP utilities: fetch wrapper, error parsing, response unwrapping.
 */

import {
  VectorDBError,
  NotFoundError,
  AlreadyExistsError,
  DimensionMismatchError,
  AuthenticationError,
  RateLimitError,
  ValidationError,
} from "./errors.js";
import type { ApiResponse } from "./types.js";

const STATUS_MAP: Record<number, new (msg: string, code: number) => VectorDBError> = {
  401: AuthenticationError,
  403: AuthenticationError,
  404: NotFoundError,
  409: AlreadyExistsError,
  422: ValidationError,
  429: RateLimitError,
};

export function raiseForResponse(code: number, message: string): never {
  if (code === 400 && message.toLowerCase().includes("dimension")) {
    throw new DimensionMismatchError(message, code);
  }
  const Cls = STATUS_MAP[code];
  if (Cls) throw new Cls(message, code);
  throw new VectorDBError(message, code);
}

export function unwrap<T>(body: ApiResponse<T>): T {
  if (body.status === "error") {
    const err = body.error ?? { code: 0, message: "Unknown error" };
    raiseForResponse(err.code, err.message);
  }
  return body.data as T;
}

export type FetchFn = (url: string, init?: RequestInit) => Promise<Response>;

export class HttpClient {
  private readonly baseUrl: string;
  private readonly headers: Record<string, string>;
  private readonly fetchFn: FetchFn;

  constructor(baseUrl: string, apiKey: string, fetchFn?: FetchFn) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.headers = {
      "x-api-key": apiKey,
      "Content-Type": "application/json",
      Accept: "application/json",
    };
    this.fetchFn = fetchFn ?? (globalThis.fetch as FetchFn);
  }

  async request<T>(
    method: string,
    path: string,
    options: { body?: unknown; params?: Record<string, string> } = {}
  ): Promise<T> {
    let url = `${this.baseUrl}${path}`;
    if (options.params && Object.keys(options.params).length > 0) {
      const qs = new URLSearchParams(options.params).toString();
      url = `${url}?${qs}`;
    }

    const init: RequestInit = {
      method,
      headers: this.headers,
    };
    if (options.body !== undefined) {
      init.body = JSON.stringify(options.body);
    }

    const resp = await this.fetchFn(url, init);
    const body = (await resp.json()) as ApiResponse<T>;

    if (!resp.ok) {
      // HTTP-level failure (e.g. 401 from middleware)
      const msg = body?.error?.message ?? `HTTP ${resp.status}`;
      raiseForResponse(resp.status, msg);
    }

    return unwrap(body);
  }

  /**
   * Send a multipart/form-data POST request (used for file uploads).
   *
   * @param path - API endpoint path.
   * @param fields - Plain text form fields (key-value pairs).
   * @param file - The file content as a Blob or Buffer.
   * @param filename - The filename to attach.
   */
  async requestMultipart<T>(
    path: string,
    fields: Record<string, string>,
    file: Blob | Buffer,
    filename: string
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;

    const formData = new FormData();
    for (const [key, value] of Object.entries(fields)) {
      formData.append(key, value);
    }
    // Convert Buffer to Blob if needed (Node.js compatibility)
    const blob =
      file instanceof Blob ? file : new Blob([file], { type: "text/plain" });
    formData.append("file", blob, filename);

    // Use all headers except Content-Type — the browser/runtime sets the
    // correct multipart boundary automatically.
    const { "Content-Type": _, ...headersWithoutCT } = this.headers;

    const init: RequestInit = {
      method: "POST",
      headers: headersWithoutCT,
      body: formData,
    };

    const resp = await this.fetchFn(url, init);
    const body = (await resp.json()) as ApiResponse<T>;

    if (!resp.ok) {
      const msg = body?.error?.message ?? `HTTP ${resp.status}`;
      raiseForResponse(resp.status, msg);
    }

    return unwrap(body);
  }
}

File: sdk/python/vectordb_client/_http.py
"""Shared HTTP utilities: error parsing and response handling."""

from __future__ import annotations
from typing import Any

from vectordb_client.exceptions import (
    VectorDBError,
    NotFoundError,
    AlreadyExistsError,
    DimensionMismatchError,
    AuthenticationError,
    RateLimitError,
    ValidationError,
)

_STATUS_MAP = {
    400: VectorDBError,
    401: AuthenticationError,
    403: AuthenticationError,
    404: NotFoundError,
    409: AlreadyExistsError,
    422: ValidationError,
    429: RateLimitError,
}


def _raise_for_response(status_code: int, body: dict) -> None:
    """Raise the appropriate SDK exception based on HTTP/body status + error body.

    The VectorDB API returns HTTP 200 for all responses, embedding the real
    error code in ``body["error"]["code"]``.  This function accepts either
    the HTTP-level status (for transport errors) or the body-level code.
    """
    error = body.get("error") or {}
    # Prefer the body-embedded error code over the HTTP status code.
    code = error.get("code", status_code)
    message = error.get("message", f"HTTP {code}")
    detail = error.get("detail", "")
    if detail:
        message = f"{message}: {detail}"

    # Special-case dimension mismatch (comes as 400)
    if code == 400 and "dimension" in message.lower():
        raise DimensionMismatchError(message, status_code=code)

    exc_cls = _STATUS_MAP.get(code, VectorDBError)
    raise exc_cls(message, status_code=code)


def _check_body_error(body: dict) -> None:
    """Check the body envelope and raise if status == 'error'."""
    if body.get("status") == "error":
        _raise_for_response(0, body)


def _unwrap(body: dict) -> Any:
    """Return the `data` payload from a standard API response envelope."""
    _check_body_error(body)
    return body.get("data", body)
